libraries/AP_Follow/AP_Follow.cpp

```diff
@@ -0,0 +1,281 @@
+
+    // return latest position estimate
+    loc = last_loc;
+    return true;
+}
+
+// get distance vector to target (in meters) and target's velocity all in NED frame
+bool AP_Follow::get_target_dist_and_vel_ned(Vector3f &dist, Vector3f &dist_with_offs, Vector3f &vel)
+{
+    // get our location
+    Location current_loc;
+    if (!_ahrs.get_position(current_loc)) {
+         return false;
+     }
+
+    // get target location and velocity
+    Location target_loc;
+    Vector3f veh_vel;
+    if (!get_target_location_and_velocity(target_loc, veh_vel)) {
+        return false;
+    }
+
+    // calculate difference
+    Vector3f dist_vec = location_3d_diff_NED(current_loc, target_loc);
+
+    // fail if too far
+    if (is_positive(_dist_max.get()) && (dist_vec.length() > _dist_max)) {
+        return false;
+    }
+
+    // initialise offsets from distance vector if required
+    init_offsets_if_required(dist_vec);
+
+    // get offsets
+    Vector3f offsets;
+    if (!get_offsets_ned(offsets)) {
+        return false;
+    }
+
+    // return results
+    dist = dist_vec;
+    dist_with_offs = dist_vec + offsets;
+    vel = veh_vel;
+    return true;
+}
+
+// get target's heading in degrees (0 = north, 90 = east)
+bool AP_Follow::get_target_heading(float &heading) const
+{
+    // exit immediately if not enabled
+    if (!_enabled) {
+        return false;
+    }
+
+    // check for timeout
+    if ((_last_heading_update_ms == 0) || (AP_HAL::millis() - _last_heading_update_ms > AP_FOLLOW_TIMEOUT_MS)) {
+        return false;
+    }
+
+    // return latest heading estimate
+    heading = _target_heading;
+    return true;
+}
+
+// handle mavlink DISTANCE_SENSOR messages
+void AP_Follow::handle_msg(const mavlink_message_t &msg)
+{
+    // exit immediately if not enabled
+    if (!_enabled) {
+        return;
+    }
+
+    // skip our own messages
+    if (msg.sysid == mavlink_system.sysid) {
+        return;
+    }
+
+    // skip message if not from our target
+    if ((_sysid != 0) && (msg.sysid != _sysid)) {
+        return;
+    }
+
+    // decode global-position-int message
+    if (msg.msgid == MAVLINK_MSG_ID_GLOBAL_POSITION_INT) {
+        uint32_t now = AP_HAL::millis();
+        // decode message
+        mavlink_global_position_int_t packet;
+        mavlink_msg_global_position_int_decode(&msg, &packet);
+        _target_location.lat = packet.lat;
+        _target_location.lng = packet.lon;
+        _target_location.alt = packet.alt/10;       // convert millimeters to cm
+        _target_velocity_ned.x = packet.vx/100.0f;  // velocity north
+        _target_velocity_ned.y = packet.vy/100.0f;  // velocity east
+        _target_velocity_ned.z = packet.vz/100.0f;  // velocity down
+        _last_location_update_ms = now;
+        if (packet.hdg <= 36000) {                  // heading (UINT16_MAX if unknown)
+            _target_heading = packet.hdg / 100.0f;  // convert centi-degrees to degrees
```

Comment:
Use `* 0.01f` instead?

Reply:
@amilcarlucas, thanks for the review, fixed!